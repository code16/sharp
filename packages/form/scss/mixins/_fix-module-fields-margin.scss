@mixin fix-module-fields-margin {
  .SharpGrid__row:last-child {
    .SharpGrid__col {
      margin-bottom: 0;
    }
    @include media-breakpoint-down(sm) {
      .SharpGrid__col.col-12:not(:last-of-type) {
        margin-bottom: $form-item-margin-bottom;
      }
    }
  }
}