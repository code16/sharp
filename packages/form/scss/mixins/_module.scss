@mixin module-contextual-close-button {
  .SharpModule {
    .remove-button, [class*=remove-button] {
      display:none;
    }
    &__inner {
      position:relative;
    }
    &__close-button {
      display:inline-block;
    }
  }
}