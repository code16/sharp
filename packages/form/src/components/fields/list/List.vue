<template>
    <div class="SharpList" :class="classes">
        <template v-if="showSortButton">
            <button type="button"
                    class="SharpButton SharpButton--ghost SharpList__sort-button"
                    :class="{'SharpButton--active':dragActive}"
                    :data-inactive-text="l('form.list.sort_button.inactive')"
                    :data-active-text="l('form.list.sort_button.active')"
                    @click="toggleDrag">
                <svg class="SharpButton__icon" width='24' height='22' viewBox='0 0 24 22' fill-rule='evenodd'>
                    <path d='M20 14V0h-4v14h-4l6 8 6-8zM4 8v14h4V8h4L6 0 0 8z'></path>
                </svg>
            </button>
        </template>
        <Draggable :options="dragOptions" :list="list" ref="draggable">
            <transition-group name="expand" tag="div">
                <div v-for="(listItemData, i) in list" :key="listItemData[indexSymbol]"
                    class="SharpList__item"
                    :class="{'SharpList__item--collapsed':dragActive}"
                >
                    <div class="SharpModule__inner">
                        <div class="SharpModule__content">

                            <template v-if="dragActive && collapsedItemTemplate">
                                <TemplateRenderer
                                    name="CollapsedItem"
                                    :template="collapsedItemTemplate"
                                    :template-data="collapsedItemData(listItemData)"
                                />
                            </template>

                            <template v-else>
                                <ListItem :layout="fieldLayout.item" :error-identifier="i" v-slot="{ fieldLayout }">
                                    <FieldDisplay
                                        :field-key="fieldLayout.key"
                                        :context-fields="transformedFields(i)"
                                        :context-data="listItemData"
                                        :error-identifier="fieldLayout.key"
                                        :config-identifier="fieldLayout.key"
                                        :update-data="update(i)"
                                        :locale="listItemData._fieldsLocale[fieldLayout.key]"
                                        :read-only="isReadOnly"
                                        @locale-change="(key, value)=>updateLocale(i, key, value)"
                                    />
                                </ListItem>
                                <template v-if="showRemoveButton">
                                    <button class="SharpButton SharpButton--danger SharpButton--sm mt-3" @click="remove(i)">
                                        {{ l('form.list.remove_button') }}
                                    </button>
                                </template>
                            </template>

                            <!-- Full size div use to handle the item when drag n drop (c.f draggable options) -->
                            <template v-if="dragActive">
                                <div class="SharpList__overlay-handle"></div>
                            </template>
                        </div>
                    </div>
                    <template v-if="showInsertButton && i < list.length-1">
                        <div class="SharpList__new-item-zone">
                            <button class="SharpButton SharpButton--sm" @click="insertNewItem(i, $event)">{{ l('form.list.insert_button') }}</button>
                        </div>
                    </template>
                </div>
            </transition-group>
            <template v-if="showAddButton" v-slot:footer>
                <button class="SharpButton SharpButton--ghost SharpList__add-button" type="button" @click="add" :key="-1">
                    {{addText}}
                </button>
            </template>
        </Draggable>
        <template v-if="readOnly && !list.length">
            <em class="SharpList__empty-alert">{{l('form.list.empty')}}</em>
        </template>
    </div>
</template>
<script>
    import Draggable from 'vuedraggable';
    import { TemplateRenderer } from 'sharp/components';
    import { Localization } from 'sharp/mixins';
    import ListItem from './ListItem';

    import localize from '../../../mixins/localize/form';
    import { transformFields, getDependantFieldsResetData, fieldEmptyValue } from "../../../util";

    export default {
        name: 'SharpList',

        inject: ['$form'],

        mixins: [ Localization,  localize('itemFields') ],

        components: {
            Draggable,
            ListItem,
            TemplateRenderer,
        },

        props: {
            fieldKey: String,
            fieldLayout: Object,
            value: Array,

            addable: {
                type:Boolean,
                default: true
            },
            sortable: {
                type: Boolean,
                default: false
            },
            removable: {
                type: Boolean,
                default: false
            },
            addText: {
                type:String,
                default:'Ajouter un élément'
            },
            itemFields: {
                type: Object,
                required: true,
            },
            collapsedItemTemplate: String,
            maxItemCount: Number,

            itemIdAttribute: String,
            readOnly:Boolean,
            locale:String
        },
        data() {
            return {
                list: [],
                itemFieldsLocale: [],
                dragActive: false,
                lastIndex: 0
            }
        },
        computed: {
            classes() {
                return {
                    'SharpList--can-sort': this.showSortButton,
                }
            },
            disabled() {
                return this.readOnly || this.dragActive;
            },
            dragOptions() {
                return {
                    disabled:!this.dragActive,
                    handle: '.SharpList__overlay-handle',
                };
            },
            showAddButton() {
                return this.addable &&
                    (this.list.length < this.maxItemCount || !this.maxItemCount) &&
                    !this.disabled;
            },
            showInsertButton() {
                return this.showAddButton && this.sortable && !this.disabled;
            },
            showSortButton() {
                return !this.hasPendingActions && this.sortable && this.list.length > 1;
            },
            showRemoveButton() {
                return this.removable && !this.disabled;
            },
            dragIndexSymbol() {
                return Symbol('dragIndex');
            },
            indexSymbol() {
                return Symbol('index');
            },
            hasPendingActions() {
                return this.$form?.hasUploadingFields(this.fieldKey);
            },
            isReadOnly() {
                return this.readOnly || this.dragActive;
            }
        },
        methods: {
            itemData(item) {
                const { id, _fieldsLocale, ...data } = item;
                return data;
            },
            transformedFields(i) {
                const item = this.list[i];
                const data = this.itemData(item);
                return transformFields(this.itemFields, data);
            },
            indexedList() {
                return (this.value||[]).map((v,i) => this.withLocale({
                    [this.indexSymbol]:i, ...v
                }));
            },
            createItem() {
                return Object.entries(this.itemFields).reduce((res, [key, field]) => {
                    res[key] = fieldEmptyValue(field.type);
                    return res;
                }, this.withLocale({
                    [this.itemIdAttribute]:null,
                    [this.indexSymbol]:this.lastIndex++
                }));
            },
            insertNewItem(i, $event) {
                $event.target && $event.target.blur();
                this.list.splice(i+1, 0, this.createItem());
            },
            add() {
                this.list.push(this.createItem());
            },
            remove(i) {
                this.list.splice(i,1);
            },
            update(i) {
                return (key, value, { forced } = {}) => {
                    const item = { ...this.list[i] };
                    const data = {
                        ...(!forced ? getDependantFieldsResetData(this.itemFields, key, () =>
                            this.fieldLocalizedValue(key, null, item, item._fieldsLocale)
                        ) : null),
                        [key]: this.fieldLocalizedValue(key, value, item, item._fieldsLocale),
                    };
                    Object.assign(this.list[i], data);
                }
            },
            updateLocale(i, key, value) {
                this.$set(this.list[i]._fieldsLocale, key, value);
            },
            collapsedItemData(itemData) {
                return {$index:itemData[this.dragIndexSymbol], ...itemData};
            },
            toggleDrag() {
                this.dragActive = !this.dragActive;
                this.list.forEach((item,i) => item[this.dragIndexSymbol] = i);
            },
            withLocale(item) {
                return {
                    ...item, _fieldsLocale: this.defaultFieldLocaleMap({
                        fields: this.itemFields,
                        locales: this.$form.locales
                    })
                };
            },

            initList() {
                this.list = this.indexedList();
                this.lastIndex = this.list.length;
                // make value === list, to update changes
                this.$emit('input', this.list);
            },
        },
        created() {
            this.localized = this.$form.localized;
            this.initList();
        },
    }
</script>
