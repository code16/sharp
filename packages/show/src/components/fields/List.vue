<template>
    <FieldLayout class="ShowListField" :class="classes" :label="label">
        <div class="ShowListField__content">
            <template v-if="isEmpty">
                <em class="ShowListField__empty text-muted">{{ lang('show.list.empty') }}</em>
            </template>
            <template v-else>
                <div class="ShowListField__list">
                    <template v-for="item in value">
                        <div class="ShowListField__item">
                            <Grid class="ShowListField__fields-grid" :rows="layout.item" v-slot="{ itemLayout:fieldLayout }">
                                <template v-if="fieldOptions(fieldLayout)">
                                    <ShowField
                                        :options="fieldOptions(fieldLayout)"
                                        :value="fieldValue(item, fieldLayout)"
                                        :config-identifier="fieldLayout.key"
                                    />
                                </template>
                                <template v-else>
                                    <UnknownField :name="fieldLayout.key" />
                                </template>
                            </Grid>
                        </div>
                    </template>
                </div>
            </template>
        </div>
    </FieldLayout>
</template>

<script>
    import { lang } from 'sharp';
    import { Grid } from 'sharp-ui';
    import { UnknownField } from 'sharp/components';
    import { syncVisibility } from "../../util/fields/visiblity";
    import FieldLayout from "../FieldLayout";

    export default {
        components: {
            Grid,
            UnknownField,
            FieldLayout,
        },
        props: {
            value: Array,
            itemFields: {
                type: Object,
                required: true,
            },
            layout: Object,
            label: String,
            emptyVisible: Boolean,
        },
        computed: {
            isEmpty() {
                return !this.value || this.value.length === 0;
            },
            isVisible() {
                return !this.isEmpty || this.emptyVisible;
            },
            classes() {
                return {
                    'ShowListField--empty': this.isEmpty,
                }
            },
            fileFieldsCollapsed() {
                const fileValues = this.allValuesOfType('file');
                return fileValues.every(value => value && !value.thumbnail);
            },
        },
        methods: {
            lang,
            fieldOptions(layout) {
                const options = this.itemFields
                    ? { ...this.itemFields[layout.key] }
                    : null;
                if(!options) {
                    console.error(`Show list field: unknown field "${layout.key}"`);
                }
                if(options.type === 'file') {
                    options.collapsed = this.fileFieldsCollapsed;
                }
                return options;
            },
            fieldValue(item, layout) {
                return item ? item[layout.key] : null;
            },
            allValuesOfType(fieldType) {
                return (this.value || []).reduce((res, item) => [
                    ...res,
                    ...Object.entries(item)
                        .filter(([key]) => {
                            const options = this.itemFields[key];
                            return options && options.type === fieldType;
                        })
                        .map(([key, value]) => value)
                ], []);
            },
        },
        created() {
            syncVisibility(this, () => this.isVisible);
        }
    }
</script>