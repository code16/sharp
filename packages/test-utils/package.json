{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}