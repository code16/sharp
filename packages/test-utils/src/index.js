export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}