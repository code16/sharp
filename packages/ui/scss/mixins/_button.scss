@mixin button-theme($bg-color, $border-color, $font-color, $hover-bg-color, $icon-color) {
  background-color: $bg-color;
  border-width: $button-border-width;
  border-style: solid;
  border-color: $border-color;
  color: $font-color;

  &:focus,
  &:hover {
    background-color: $hover-bg-color;
  }

  &:focus {
    border: $button-border-width solid $ui-02;
  }

  &:disabled:hover,
  &:disabled:focus {
    background-color: $bg-color;
  }

  &:active {
    background-color: darken($bg-color, 20%);
  }

  svg {
    fill: $icon-color;
  }
}