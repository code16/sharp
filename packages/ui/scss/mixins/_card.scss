@mixin app-status($status) {
  @include reset;
  @include font-size('12');
  display: none;
  align-items: center;

  &::before {
    content: ' ';
    height: 1rem;
    width: 1rem;
    border-radius: 100%;
    margin-right: .25rem;
  }

  @if ($status == 'running') {
    &::before {
      background-color: $support-02;
    }
  }

  @else if ($status == 'not-running') {
    &::before {
      background-color: $support-01;
    }
  }
}
