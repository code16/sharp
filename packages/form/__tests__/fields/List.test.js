import Vue from 'vue';
import List from '../../src/components/fields/list/List.vue';

import { MockInjections, MockTransitions, MockI18n, QueryComponent } from '@sharp/test-utils';
import { ErrorNode } from 'sharp/mixins';

import { mount } from '@vue/test-utils';

describe('list-field', () => {
    Vue.use(MockTransitions);
    Vue.use(MockI18n);
    Vue.use(QueryComponent);

   const fieldDisplayMock = Vue.component('FieldDisplay',{
       name: 'SharpFieldContainer',
       inheritAttrs: false,
       mixins: [ ErrorNode ],
       render:h=>h('div',' FIELD DISPLAY MOCK ')
   });

    beforeEach(()=>{
        document.body.innerHTML = `    
            <div id="app">
                <sharp-list 
                    :value="value" 
                    :field-layout="{ 
                        item:[
                            [ {key:'name'} ]
                        ]
                    }"
                    :item-fields="itemFields || { name: { type:'text' } }"
                    :addable="addable" 
                    :sortable="sortable"
                    :removable="removable"
                    :collapsed-item-template="'<span> {{name}} </span>'"
                    :max-item-count="5"
                    item-id-attribute="id"
                    :read-only="readOnly"
                    locale="fr"
                    @input="inputEmitted" 
                />
            </div>
        `;
        MockI18n.mockLangFunction();
    });

    test('can mount empty list field', async () => {
        await createVm({
            propsData: {
                addable:true, removable:true, sortable:true
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount empty "read only" list field', async () => {
        await createVm({
            propsData: {
                readOnly: true,
                addable:true, removable:true, sortable:true
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount filled list field', async () => {
        await createVm({
            propsData: {
                addable:true, removable:true, sortable:true
            },
            data:()=>({
                value: [{id:0, name:'Antoine'}, {id:1, name:'Samuel'}]
            })
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount "read only" filled list field', async () => {
        await createVm({
            propsData: {
                readOnly:true,
                addable: true, removable: true, sortable: true
            },
            data:()=>({
                value: [{id:0, name:'Antoine'}]
            })
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount collapsed list field', async () => {
        let $list = await createVm({
            propsData: {
                addable: true, removable: true, sortable: true
            },
            data:()=>({
                value: [{id:0, name:'Antoine'}]
            }),
        });

        expect.assertions(1);

        $list.dragActive = true;

        await Vue.nextTick();
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount non addable, non removable, non sortable list field', async () => {
        await createVm({
            propsData: {
                addable: false, removable: false, sortable: false
            },
            data:()=>({
                value: [{id:0, name:'Antoine'},{ id:1, name:'Samuel' }]
            }),
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount with full list', async () => {
        await createVm({
            propsData: {
                addable:true, removable:true, sortable: true
            },
            data:()=>({
                value: [{id:0, name:'Antoine'},{id:1, name:'Samuel'},{id:2, name:'Solène'},{id:3, name:'Georges'},{id:4, name:'Gérard'}]
            }),
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('emit input on init to have list and value equals by reference (sync changes)', async () => {
        let inputEmitted = jest.fn();

        let $list = await createVm({
            methods: {
                inputEmitted
            }
        });

        expect(inputEmitted).toHaveBeenCalledTimes(1);
        expect(inputEmitted).toHaveBeenCalledWith($list.list);
    });

    test('create appropriate item', async () => {
        let $list = await createVm();

        let item = $list.createItem();

        expect(item).toMatchObject({ id: null, name: null });
    });

    test('items have correct indexes', async () => {
        let $list = await createVm({
            data:()=>({
                value:[{id:0, name:'Antoine'},{id:1, name:'Samuel'},{id:2, name:'Solène'}]
            })
        });

        let { list, indexSymbol } = $list;

        expect(list).toMatchObject([
            { [indexSymbol]:0 },{ [indexSymbol]:1 },{ [indexSymbol]:2 }
        ]);

        $list.add();

        expect(list).toMatchObject([
            { [indexSymbol]:0 },{ [indexSymbol]:1 },{ [indexSymbol]:2 },{ [indexSymbol]:3 }
        ]);
    });

    test('items have correct drag indexes', async () => {
        let $list = await createVm({
            data:()=>({
                value:[{id:0, name:'Antoine'},{id:1, name:'Samuel'},{id:2, name:'Solène'}]
            })
        });

        $list.toggleDrag();

        let { list, dragIndexSymbol } = $list;

        expect(list).toMatchObject([
            { [dragIndexSymbol]:0 }, { [dragIndexSymbol]:1 }, { [dragIndexSymbol]:2 }
        ]);
    });

    test('remove item correctly', async () => {
        let $list = await createVm({
            data:()=>({
                value:[{id:0, name:'Antoine'},{id:1, name:'Samuel'},{id:2, name:'Solène'}]
            })
        });

        $list.remove(1);

        let { list, indexSymbol } = $list;

        expect(list).toMatchObject([
            {id:0, name:'Antoine', [indexSymbol]:0 },{id:2, name:'Solène', [indexSymbol]:2}
        ])
    });

    test('expose appropriate collapsed item template props data', async () => {
        let $list = await createVm();

        let { dragIndexSymbol } = $list;
        let data = $list.collapsedItemData({ id:1, name:'Samuel', [dragIndexSymbol]: 2 });

        expect(data).toMatchObject({
            $index: 2,
            id: 1,
            name: 'Samuel'
        });
    });

    test('expose appropriate props to field-display', () => {
        let wrapper = mount(List,{
            provide: MockInjections.provide,
            propsData: {
                value: [ { id:0, name:'myName' }],
                fieldLayout: {
                    item:[
                        [ {key:'name'} ]
                    ]
                },
                itemFields: { name: { type:'text' } },
                itemIdAttribute:"id",
            },
            methods: {
                update: jest.fn(i => `update ${i}`)
            }
        });
        let $field = wrapper.find(fieldDisplayMock);
        expect($field.vm.$options.propsData).toMatchObject({
            errorIdentifier: 'name'
        });
        expect($field.vm.$attrs).toEqual({
            'field-key': 'name',
            'context-fields': { name: { type:'text', readOnly:false, } },
            'context-data': expect.objectContaining({ id: 0, name: 'myName', _fieldsLocale: {} }),
            'config-identifier': 'name',
            'update-data': 'update 0',
            'read-only': false,
        });
    });

    test('update data properly', async () => {
        const provided = MockInjections.provide();
        provided.$form.localized = true;
        provided.$form.locales = ['fr', 'en'];
        let $list = await createVm({
            provide: provided,
            propsData: {
                itemFields: {
                    name: { key:'name', type:'text' },
                    localizedField: { key:'localizedField', type:'text', localized: true }
                },
            },

            data:()=>({
                value:[{id:0, name:'Antoine'},{id:1, name:'Samuel'},{id:2, name:'Solène'}]
            })
        }, null, { mockInjections:false });

        let updateFn = $list.update(1);
        $list.fieldLocalizedValue = jest.fn(()=>'fieldLocalizedValue');

        updateFn('name','George');
        expect($list.list[1]).toMatchObject({ id:1, name:'fieldLocalizedValue' });
        expect($list.fieldLocalizedValue).toHaveBeenCalledWith(
            'name', 'George',
            expect.objectContaining({id:1, name:'Samuel'}),
            { localizedField:'fr' }
        );

        jest.resetAllMocks();

        updateFn('localizedField', 'aaa');
        expect($list.list[1]).toMatchObject({ id:1, name:'fieldLocalizedValue', localizedField:undefined });
        expect($list.fieldLocalizedValue).toHaveBeenCalledWith(
            'localizedField', 'aaa',
            expect.objectContaining({id:1, name:'fieldLocalizedValue' }),
            { localizedField:'fr' }
        );
    });

    test('insert item properly', async () => {
        let $list = await createVm({
            data:()=>({
                value:[{id:0, name:'Antoine'},{id:2, name:'Solène'}]
            })
        });

        $list.insertNewItem(0, {});

        expect($list.list).toMatchObject([
            {id:0, name:'Antoine'},{id: null, name: null},{id:2, name:'Solène'}
        ])
    });

    test('have proper field identifier', async () => {
        let $list = await createVm({
            data:()=>({
                value:[{id:0, name:'Antoine'}, {id:1, name:'Samuel'}]
            })
        });

        let identifiers = $list.$findDeepChildren('SharpFieldContainer').map(fc => fc.mergedErrorIdentifier);

        expect(identifiers).toEqual(expect.arrayContaining([ '0.name', '1.name' ]));
    });

    test('has localize mixin with right fieldsProps', async () => {
        let $list = await createVm();
        expect($list.$options._localizedForm).toBe('itemFields');
    });
});

async function createVm(customOptions={}, mock, { mockInjections=true }={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [
            mockInjections ? MockInjections: {},
            customOptions
        ],

        components: {
            'sharp-list':mock||List
        },

        props:['readOnly', 'addable', 'sortable', 'removable', 'itemFields'],

        'extends': {
            data:() => ({
                value: null
            }),
            methods: {
                inputEmitted: ()=>{}
            }
        }

    });

    await Vue.nextTick();

    return vm.$children[0];
}