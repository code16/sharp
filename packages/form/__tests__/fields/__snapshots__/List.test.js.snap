// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list-field can mount "read only" filled list field 1`] = `
<div id="app">
	<div class="SharpList">
		<!---->
		<div>
			<div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<!---->
							<!---->
						</div>
					</div>
					<!---->
				</div>
			</div>
		</div>
		<!---->
	</div>
</div>
`;

exports[`list-field can mount collapsed list field 1`] = `
<div id="app">
	<div class="SharpList">
		<!---->
		<div>
			<div>
				<div class="SharpList__item SharpList__item--collapsed">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpTemplate">
								<span> Antoine </span>
							</div>
							<div class="SharpList__overlay-handle"></div>
						</div>
					</div>
					<!---->
				</div>
			</div>
		</div>
		<!---->
	</div>
</div>
`;

exports[`list-field can mount empty "read only" list field 1`] = `
<div id="app">
	<div class="SharpList">
		<!---->
		<div>
			<div></div>
		</div>
		<em class="SharpList__empty-alert">{{ form.list.empty }}</em>
	</div>
</div>
`;

exports[`list-field can mount empty list field 1`] = `
<div id="app">
	<div class="SharpList">
		<!---->
		<div>
			<div></div>
			<button type="button" class="SharpButton SharpButton--ghost SharpList__add-button">
				Ajouter un élément
			</button>
		</div>
		<!---->
	</div>
</div>
`;

exports[`list-field can mount filled list field 1`] = `
<div id="app">
	<div class="SharpList SharpList--can-sort">
		<button type="button" data-inactive-text="{{ form.list.sort_button.inactive }}" data-active-text="{{ form.list.sort_button.active }}" class="SharpButton SharpButton--ghost SharpList__sort-button">
			<svg width="24" height="22" viewBox="0 0 24 22" fill-rule="evenodd" class="SharpButton__icon">
				<path d="M20 14V0h-4v14h-4l6 8 6-8zM4 8v14h4V8h4L6 0 0 8z"></path>
			</svg>
		</button>
		<div>
			<div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<button class="SharpButton SharpButton--danger SharpButton--sm mt-3">
								{{ form.list.remove_button }}
							</button>
							<!---->
						</div>
					</div>
					<div class="SharpList__new-item-zone">
						<button class="SharpButton SharpButton--sm">{{ form.list.insert_button }}</button>
					</div>
				</div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<button class="SharpButton SharpButton--danger SharpButton--sm mt-3">
								{{ form.list.remove_button }}
							</button>
							<!---->
						</div>
					</div>
					<!---->
				</div>
			</div>
			<button type="button" class="SharpButton SharpButton--ghost SharpList__add-button">
				Ajouter un élément
			</button>
		</div>
		<!---->
	</div>
</div>
`;

exports[`list-field can mount non addable, non removable, non sortable list field 1`] = `
<div id="app">
	<div class="SharpList">
		<!---->
		<div>
			<div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<!---->
							<!---->
						</div>
					</div>
					<!---->
				</div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<!---->
							<!---->
						</div>
					</div>
					<!---->
				</div>
			</div>
		</div>
		<!---->
	</div>
</div>
`;

exports[`list-field can mount with full list 1`] = `
<div id="app">
	<div class="SharpList SharpList--can-sort">
		<button type="button" data-inactive-text="{{ form.list.sort_button.inactive }}" data-active-text="{{ form.list.sort_button.active }}" class="SharpButton SharpButton--ghost SharpList__sort-button">
			<svg width="24" height="22" viewBox="0 0 24 22" fill-rule="evenodd" class="SharpButton__icon">
				<path d="M20 14V0h-4v14h-4l6 8 6-8zM4 8v14h4V8h4L6 0 0 8z"></path>
			</svg>
		</button>
		<div>
			<div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<button class="SharpButton SharpButton--danger SharpButton--sm mt-3">
								{{ form.list.remove_button }}
							</button>
							<!---->
						</div>
					</div>
					<!---->
				</div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<button class="SharpButton SharpButton--danger SharpButton--sm mt-3">
								{{ form.list.remove_button }}
							</button>
							<!---->
						</div>
					</div>
					<!---->
				</div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<button class="SharpButton SharpButton--danger SharpButton--sm mt-3">
								{{ form.list.remove_button }}
							</button>
							<!---->
						</div>
					</div>
					<!---->
				</div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<button class="SharpButton SharpButton--danger SharpButton--sm mt-3">
								{{ form.list.remove_button }}
							</button>
							<!---->
						</div>
					</div>
					<!---->
				</div>
				<div class="SharpList__item">
					<div class="SharpModule__inner">
						<div class="SharpModule__content">
							<div class="SharpGrid">
								<div class="SharpGrid__row row">
									<div class="SharpGrid__col col-md">
										<div> FIELD DISPLAY MOCK </div>
									</div>
								</div>
							</div>
							<button class="SharpButton SharpButton--danger SharpButton--sm mt-3">
								{{ form.list.remove_button }}
							</button>
							<!---->
						</div>
					</div>
					<!---->
				</div>
			</div>
		</div>
		<!---->
	</div>
</div>
`;
